### ISSUE TITLE
Filesystem access not working for mounts outside of home

### ISSUE THREAD

**Issue by JaneSmith** [tchx84/Flatseal#709]
Hello. I've been struggling a bit with the VS Code Flatpak today...

**1. I've used Flatseal to disable host and home filesystem access, and added access to the following directory:**

`/var/home/username/Projects`

**This works just fine.** I'm able to open files and folders in the ~/Projects directory.

Note that I'm testing this using drag+drop from my file manager into VS Code, and that seems to load the filepaths directly. If I use VS Code's open dialog, it seems to use portals, which end up giving me a /run/user/1000/doc/ path which is its own can of worms...

**2. I then used Flatseal to add a second directory:**

`/var/mnt/Data/Projects`

**This one doesn't work.** If I drag+drop anything from this directory into VS Code, the app says it can't find the file.

Note that /var/mnt/Data here is a separate hard drive, permanently mounted via fstab.

If I use a terminal in the app's sandbox, I can see that there is no /var/mnt/ directory in the sandbox. I tried looking around to see if it had been mounted anywhere else, and couldn't find it — and searching online, it seems Flatpak is supposed to always use the same mount path for simplicity.

**3. I then tried putting a symlink into ~/Projects, like so:**

`/var/home/username/Projects/Test` -> `/var/mnt/Data/Projects/Test` (symlink)

**Again, this doesn't work.** I can drag+drop files from Projects into VS Code just fine, but anything within the Test subfolder doesn't work.

If I use a terminal in the app's sandbox, I can see the Test symlink, but I can't follow it — because /var/mnt/ doesn't exist in the sandbox.

**4. I then tried a bind mount instead of a symlink, like so:**

`/var/home/username/Projects/Test` -> `/var/mnt/Data/Projects/Test` (bind mount)

**This one works fine.**

***

I'm left feeling quite confused. As far as I can tell, I can add filesystem access to anything under home, but something mounted outside of home won't work. I thought that perhaps `/var/` might be a reserved directory (even though Flatseal didn't report any errors), but the Flatpak docs seem to suggest that you can override `/var/`.

What's going on here? If I'm doing something wrong, then why isn't Flatseal reporting any errors? If I'm not doing anything wrong, then why isn't it working as expected?

**Comment by devinrsmith** [comment]
I'm running into this as well. Fedora 41.

**Comment by devinrsmith** [comment]
I wonder if it may be related to selinux.

**Comment by rusty-snake** [comment]
Can you guys check with `flatpak run --command=sh com.visualstudio.code` whether the files are allowed by flatpak or not, and with `flatpak info -M com.visualstudio.code` whether the static permission are added correctly (i.e is is a flatseal issue, a flatpak permission issue, a documents portal issue or a vs code issue).

> I wonder if it may be related to selinux.

Check your audit log for any AVCs.

**Comment by rusty-snake** [comment]
It's also possible that zypak carries on with the wrong permissions.

**Comment by devinrsmith** [comment]
Thanks for the tip, it looks like this is helpful for debugging at least in my case. I'm using a different app, but here is the result:

```
$ flatpak run --command=sh com.jetbrains.IntelliJ-IDEA-Community
F: Not sharing "/usr/lib/jvm/java-21-zulu-openjdk-ca" with sandbox: Path "/usr" is reserved by Flatpak
[📦 com.jetbrains.IntelliJ-IDEA-Community ~]$ 
exit
```

`Path "/usr" is reserved by Flatpak` seems to be the root cause.

I created a `/var/mnt/` directory to test and that seems to share into Flatpak just fine, so I'm not sure if this is the same issue as the original poster or not. That said, I wonder if there is a workaround to allow mappings from `/usr`, or maybe an upfront warning that it's reserved?

**Comment by rusty-snake** [comment]
yeah, (host)`/usr` is impossible to allow as (flatpak)`/usr`. However you can use `host`/`host-os` to allow (host)`/usr` as (flatpak)`/run/host/usr`. See https://docs.flatpak.org/en/latest/sandbox-permissions.html#filesystem-access.

> maybe an upfront warning that it's reserved?

I think so, flatseal as a gui should inform users about that when adding reserved paths:

```
/app, /bin, /dev, /etc, /lib, /lib32, /lib64, /proc, /run/flatpak, /run/host, /sbin, /usr
```

I did not found an open issue for it, would you like to create on?

**Comment by JaneSmith** [comment]
> Can you guys check with `flatpak run --command=sh com.visualstudio.code` whether the files are allowed by flatpak or not, and with `flatpak info -M com.visualstudio.code` 

Well, I tried it again to get this information for you, and strangely it's working fine for me now. I have no idea why. I definitely added the filepath in Flatseal the same way, and I definitely checked with `--command=sh` before (as mentioned in the OP, I did try a terminal within the sandbox). I can start VS Code and drag+drop the files just fine now.

So, I don't know what happened. I can only assume that somehow the permission wasn't getting added properly before, and now it is. Sorry about that. I see a separate issue has been opened about reserved filepaths, so I'll go ahead and close this. Thanks.
